sklearn/tests/test_base.py

```diff
@@ -509,3 +516,64 @@ def test_regressormixin_score_multioutput():
            "built-in scorer 'r2' uses "
            "multioutput='uniform_average').")
     assert_warns_message(FutureWarning, msg, reg.score, X, y)
+
+
+def test_validate_X():
+    # Make sure ValueError is raised when there is a n_features mismatch
+    # between fit and predict/transform
+
+    X = [[0, 1], [2, 3]]
+
+    ss = StandardScaler().fit(X)
+    ss.transform(X)  # All good
+
+    with pytest.raises(ValueError, match="X has 3 features, but"):
+        X_more_features = [[0, 1, 4], [2, 3, 5]]
+        ss.transform(X_more_features)
+
+
+def test_n_features_in_attribute():
+    # Make sure n_features_in_ is correctly set.
+    # Note that n_features_in_ is always None for vectorizers, while for other
+    # estimators the attribute doesn't exist until fit() is called.
+    X_2 = [[0, 1], [2, 3]]
+    X_3 = [[0, 1, 4], [2, 3, 5]]
+
+    ss = StandardScaler()
+    assert not hasattr(ss, 'n_features_in_')
+    ss.fit(X_2)
+    assert ss.n_features_in_ == 2
+    ss = ss.fit(X_3)
+    assert ss.n_features_in_ == 3
+
+    dv = DictVectorizer()
+    assert dv.n_features_in_ is None
+    d = [{'foo': 1, 'bar': 2}, {'foo': 3, 'baz': 1}]
+    dv.fit(d)
+    assert dv.n_features_in_ is None
+
+    # meta estimator need specific ways of dealing with the attribute:
+    # grid search delegates this to # the best estimator
+    n_features = 4
+    X, y = make_classification(n_features=n_features)
+    gbdt = HistGradientBoostingClassifier()
+    param_grid = {'max_iter': [3, 4]}
+    gs = GridSearchCV(gbdt, param_grid)
+    assert hasattr(ss, 'n_features_in_')  # that might be a bit unintuitive
+    with pytest.raises(NotFittedError):
+        gs.n_features_in_
+    gs.fit(X, y)
+    assert gs.n_features_in_ == n_features
+
+    # pipelines delegate to the first step
```

Comment:
I personally think this belongs in test_pipeline.py, being about its specific functionality.